Mathlib/Combinatorics/Colex.lean

```diff
@@ -406,19 +407,50 @@ lemma geomSum_ofColex_strictMono (hn : 2 ≤ n) : StrictMono fun s ↦ ∑ k ∈
   exact (Nat.geomSum_lt hn <| by simpa).trans_le <| single_le_sum (fun _ _ ↦ by positivity) <|
     mem_sdiff.2 ⟨hat, has⟩
 
-/-- For finsets of naturals of naturals, the colexicographic order is equivalent to the order
-induced by the `n`-ary expansion. -/
+/-- For finsets of naturals, the colexicographic order is equivalent to the order induced by the
+`n`-ary expansion. -/
 lemma geomSum_le_geomSum_iff_toColex_le_toColex (hn : 2 ≤ n) :
     ∑ k ∈ s, n ^ k ≤ ∑ k ∈ t, n ^ k ↔ toColex s ≤ toColex t :=
   (geomSum_ofColex_strictMono hn).le_iff_le
 
-/-- For finsets of naturals of naturals, the colexicographic order is equivalent to the order
-induced by the `n`-ary expansion. -/
+/-- For finsets of naturals, the colexicographic order is equivalent to the order induced by the
+`n`-ary expansion. -/
 lemma geomSum_lt_geomSum_iff_toColex_lt_toColex (hn : 2 ≤ n) :
     ∑ i ∈ s, n ^ i < ∑ i ∈ t, n ^ i ↔ toColex s < toColex t :=
   (geomSum_ofColex_strictMono hn).lt_iff_lt
 
--- TODO: Package the above in the `n = 2` case as an order isomorphism `Colex ℕ ≃o ℕ`
+theorem geomSum_injective {n : ℕ} (hn : 2 ≤ n) :
+    Function.Injective (fun s : Finset ℕ ↦ ∑ i in s, n ^ i) := by
+  intro _ _ h
+  rwa [le_antisymm_iff, geomSum_le_geomSum_iff_toColex_le_toColex hn,
+    geomSum_le_geomSum_iff_toColex_le_toColex hn, ← le_antisymm_iff, Colex.toColex.injEq] at h
+
+theorem lt_geomSum_of_mem {a : ℕ} (hn : 2 ≤ n) (hi : a ∈ s) : a < ∑ i in s, n ^ i :=
+  (Nat.lt_pow_self hn a).trans_le <| single_le_sum (by simp) hi
+
+@[simp] theorem toFinset_bitIndices_twoPowSum (s : Finset ℕ) :
+    (∑ i in s, 2 ^ i).bitIndices.toFinset = s := by
+  simp [← (geomSum_injective rfl.le).eq_iff, List.sum_toFinset _ Nat.bitIndices_sorted.nodup]
+
+@[simp] theorem twoPowSum_toFinset_bitIndices (n : ℕ) :
```

Comment:
Please explain that new `twoPowSum`naming convention next to the explanation for `geomSum`

Reply:
where exactly do you mean?

Reply:
Module doc of `Algebra.GeomSum`

Reply:
I don't see anywhere that this fits. The (obvious) convention of using  one of `geom_sum` or `geomSum` to denote geometric sum isn't really explicitly documented in that file, and adding a description of `twoPowSum` would look strange given that it isn't referenced anywhere in that file.

Reply:
I could see this making sense in a separate PR which does some tidying up, including picking one of `geom_sum` or `geomSum` to use everywhere.

Reply:
Yeah sure